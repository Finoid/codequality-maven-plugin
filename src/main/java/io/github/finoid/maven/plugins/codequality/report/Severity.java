package io.github.finoid.maven.plugins.codequality.report;

public enum Severity {
    INFO,
    MINOR,
    MAJOR,
    CRITICAL,
    BLOCKER;

    @SuppressWarnings("EnumOrdinal")
    public boolean isHigherThanOrEqual(final Severity severity) {
        return this.ordinal() >= severity.ordinal();
    }
}
