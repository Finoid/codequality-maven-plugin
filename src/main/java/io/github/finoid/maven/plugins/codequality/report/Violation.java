package io.github.finoid.maven.plugins.codequality.report;

import lombok.Value;

@Value
public class Violation {
    String description;
    String fingerprint;
    Severity severity;
    String relativePath;
    String fullPath;
    Integer line;
    Integer columnNumber;
    String rule;
}
