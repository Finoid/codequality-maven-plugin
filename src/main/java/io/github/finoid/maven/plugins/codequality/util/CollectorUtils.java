package io.github.finoid.maven.plugins.codequality.util;

import lombok.experimental.UtilityClass;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collector;
import java.util.stream.Collectors;

@UtilityClass
public class CollectorUtils {
    /**
     * Alternative to {@link Collectors#toList} which returns an unmodifiable list.
     *
     * @param <T> the type of the elements
     * @return modifiable list
     */
    public static <T> Collector<T, ?, List<T>> toMutableList() {
        return Collectors.toCollection(ArrayList::new);
    }
}
