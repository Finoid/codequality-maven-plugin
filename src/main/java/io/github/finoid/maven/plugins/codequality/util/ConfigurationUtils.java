package io.github.finoid.maven.plugins.codequality.util;

import lombok.experimental.UtilityClass;
import org.codehaus.plexus.util.xml.Xpp3Dom;

@UtilityClass
public class ConfigurationUtils {
    public static String oneOrThrow(final Xpp3Dom configuration, final String key) {
        final Xpp3Dom[] children = configuration.getChildren(key);

        if (children.length != 1) {
            throw new IllegalStateException("Illegal configuration key provided");
        }

        return children[0].getValue();
    }
}
