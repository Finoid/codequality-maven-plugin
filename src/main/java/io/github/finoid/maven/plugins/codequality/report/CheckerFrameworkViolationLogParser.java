package io.github.finoid.maven.plugins.codequality.report;

import org.codehaus.plexus.logging.Logger;
import org.codehaus.plexus.logging.console.ConsoleLogger;
import io.github.finoid.maven.plugins.codequality.exceptions.ParseException;
import io.github.finoid.maven.plugins.codequality.step.ViolationConverter;
import io.github.finoid.maven.plugins.codequality.util.Precondition;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Singleton
public class CheckerFrameworkViolationLogParser implements ViolationLogParser {
    public static final Pattern PART_VIOLATION_PATTERN = Pattern.compile("^(.*):(\\[\\d+,\\d+\\]).*(\\[.*\\])");
    public static final Pattern VIOLATION_PATTERN =
        Pattern.compile("^(?<path>.*):\\[(?<line>\\d+),(?<column>\\d+)\\].*\\[(?<rule>.*)\\] (?<description>.*)$");

    private static final Logger LOGGER = new ConsoleLogger(1, "console");

    private final ViolationConverter violationConverter;

    @Inject
    public CheckerFrameworkViolationLogParser(final ViolationConverter violationConverter) {
        this.violationConverter = Precondition.nonNull(violationConverter, "ViolationConverter shouldn't be null");
    }

    @Override
    public List<Violation> parse(final InputStream fileInputStream) {
        final List<Violation> violations = new ArrayList<>();

        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(fileInputStream, StandardCharsets.UTF_8))) {
            String nullableLine;

            while ((nullableLine = reader.readLine()) != null) {
                if (isNotCheckerFrameworkViolationLine(nullableLine)) {
                    continue;
                }

                parseViolation(nullableLine)
                    .ifPresent(violations::add);
            }

            return violations;
        } catch (final IOException e) {
            throw new ParseException("Exception during parsing", e);
        }
    }

    private Optional<Violation> parseViolation(final String currentLine) {
        final Matcher violationMatcher = VIOLATION_PATTERN.matcher(currentLine);
        if (!violationMatcher.find()) {
            LOGGER.warn("Unexpected checker framework log. Log: " + currentLine);

            return Optional.empty();
        }

        return violationOf(currentLine);
    }

    private Optional<Violation> violationOf(final String checkerFrameworkViolationLines) {
        final Matcher violationMatcher = VIOLATION_PATTERN.matcher(checkerFrameworkViolationLines);

        if (!violationMatcher.find()) {
            LOGGER.warn("Unexpected checker framework log. Log: " + checkerFrameworkViolationLines);

            return Optional.empty();
        }

        return Optional.of(violationConverter.ofCheckerFrameworkViolationMatcher(violationMatcher));
    }

    private static boolean isNotCheckerFrameworkViolationLine(final String nullableLine) {
        final Matcher partViolationMatcher = PART_VIOLATION_PATTERN.matcher(nullableLine);

        return !partViolationMatcher.find();
    }

}
