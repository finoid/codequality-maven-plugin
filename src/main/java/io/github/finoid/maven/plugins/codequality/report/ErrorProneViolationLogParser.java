package io.github.finoid.maven.plugins.codequality.report;

import io.github.finoid.maven.plugins.codequality.exceptions.ParseException;
import io.github.finoid.maven.plugins.codequality.step.ViolationConverter;
import io.github.finoid.maven.plugins.codequality.util.Precondition;
import org.codehaus.plexus.logging.Logger;
import org.codehaus.plexus.logging.console.ConsoleLogger;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Singleton
public class ErrorProneViolationLogParser implements ViolationLogParser {
    public static final Pattern VIOLATION_PATTERN =
        Pattern.compile("^(?<path>.*):\\[(?<line>\\d+)(?:,(?<column>\\d+))?\\] \\[(?<rule>.*)\\] (?<description>.*\\s*\\(.*\\s*.*)$");

    private static final Pattern PART_VIOLATION_PATTERN = Pattern.compile("^(.*):\\[(\\d+(?:,\\d+)?)\\] \\[(.*?)\\]");
    private static final Logger LOGGER = new ConsoleLogger(1, "console");
    private static final String THIRD_LINE_PREFIX = "  Did you mean";
    private static final int READ_AHEAD_LIMIT = 8 * 1024;

    private final ViolationConverter violationConverter;

    @Inject
    public ErrorProneViolationLogParser(final ViolationConverter violationConverter) {
        this.violationConverter = Precondition.nonNull(violationConverter, "ViolationConverter shouldn't be null");
    }

    @Override
    public List<Violation> parse(final InputStream fileInputStream) {
        final List<Violation> violations = new ArrayList<>();

        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(fileInputStream, StandardCharsets.UTF_8))) {
            String nullableLine;

            while ((nullableLine = reader.readLine()) != null) {
                if (isNotErrorProneViolationLine(nullableLine)) {
                    continue;
                }

                parseViolation(nullableLine, reader)
                    .ifPresent(violations::add);
            }

            return violations;
        } catch (final IOException e) {
            throw new ParseException("Exception during parsing", e);
        }
    }

    private Optional<Violation> parseViolation(final String currentLine, final BufferedReader reader) throws IOException {
        final StringBuilder errorProneViolationBuffer = new StringBuilder();

        errorProneViolationBuffer
            .append(currentLine)
            .append(reader.readLine());

        reader.mark(READ_AHEAD_LIMIT);

        String peakAhead = reader.readLine();

        // optional third line: typically "  Did you mean ..."
        if (peakAhead.startsWith(THIRD_LINE_PREFIX)) {
            errorProneViolationBuffer.append(peakAhead);
        } else {
            // Not a valid third line, rewind so the outer loop will process it
            reader.reset();
        }

        final Matcher violationMatcher = VIOLATION_PATTERN.matcher(errorProneViolationBuffer.toString());
        if (!violationMatcher.find()) {
            LOGGER.warn("Unexpected error prone log. Log: " + errorProneViolationBuffer);

            return Optional.empty();
        }

        return violationOf(errorProneViolationBuffer.toString());
    }

    private Optional<Violation> violationOf(final String errorProneViolationLines) {
        final Matcher violationMatcher = VIOLATION_PATTERN.matcher(errorProneViolationLines);

        if (!violationMatcher.find()) {
            LOGGER.warn("Unexpected error prone log. Log: " + errorProneViolationLines);

            return Optional.empty();
        }

        return Optional.of(violationConverter.ofErrorProneViolationMatcher(violationMatcher));
    }

    private static boolean isNotErrorProneViolationLine(final String nullableLine) {
        final Matcher partViolationMatcher = PART_VIOLATION_PATTERN.matcher(nullableLine);

        return !partViolationMatcher.find();
    }
}
