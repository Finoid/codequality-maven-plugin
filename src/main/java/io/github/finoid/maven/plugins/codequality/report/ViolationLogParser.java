package io.github.finoid.maven.plugins.codequality.report;

import io.github.finoid.maven.plugins.codequality.exceptions.ParseException;

import java.io.InputStream;
import java.util.List;

public interface ViolationLogParser {

    /**
     * Parses the provided input stream to extract a list of violations from log entries.
     *
     * @param fileInputStream the input stream containing the log data to be parsed.
     * @return a list of {@link Violation} objects representing the identified violations in the logs.
     * @throws ParseException if an error occurs during the parsing process, such as invalid log formatting or issues reading from the input stream.
     */
    List<Violation> parse(final InputStream fileInputStream);

}
