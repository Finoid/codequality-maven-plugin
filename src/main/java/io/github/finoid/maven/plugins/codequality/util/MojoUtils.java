package io.github.finoid.maven.plugins.codequality.util;

import lombok.experimental.UtilityClass;
import org.apache.maven.model.Dependency;
import org.apache.maven.model.Plugin;
import org.apache.maven.plugin.descriptor.PluginDescriptor;
import org.twdata.maven.mojoexecutor.MojoExecutor;

import java.util.List;

import static org.twdata.maven.mojoexecutor.MojoExecutor.artifactId;
import static org.twdata.maven.mojoexecutor.MojoExecutor.element;
import static org.twdata.maven.mojoexecutor.MojoExecutor.groupId;
import static org.twdata.maven.mojoexecutor.MojoExecutor.name;
import static org.twdata.maven.mojoexecutor.MojoExecutor.version;

@UtilityClass
public class MojoUtils {
    @UtilityClass
    public static class ElementUtils {
        public static MojoExecutor.Element annotationProcessor(final String groupId, final String artifactId, final String version) {
            return element(name("path"),
                element(name("groupId"), groupId),
                element(name("artifactId"), artifactId),
                element(name("version"), version)
            );
        }
    }

    @UtilityClass
    public static class PluginUtils {
        public static PluginDescriptor pluginDescriptor(final String groupId, final String artifactId, final String version) {
            final PluginDescriptor descriptor = new PluginDescriptor();
            descriptor.setGroupId(groupId);
            descriptor.setArtifactId(artifactId);
            descriptor.setVersion(version);

            return descriptor;
        }

        public static Plugin pluginOfDescriptor(final PluginDescriptor descriptor) {
            return MojoExecutor.plugin(
                groupId(descriptor.getGroupId()),
                artifactId(descriptor.getArtifactId()),
                version(descriptor.getVersion())
            );
        }

        public static Plugin plugin(
            final String groupId,
            final String artifactId,
            final String version,
            final List<Dependency> dependencies
        ) {
            final Plugin plugin = new Plugin();
            plugin.setArtifactId(artifactId);
            plugin.setGroupId(groupId);
            plugin.setVersion(version);
            plugin.setDependencies(dependencies);

            return plugin;
        }
    }
}
