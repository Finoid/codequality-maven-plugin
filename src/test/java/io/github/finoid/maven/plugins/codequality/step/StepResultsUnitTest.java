package io.github.finoid.maven.plugins.codequality.step;

import io.github.finoid.maven.plugins.codequality.fixtures.ProjectStepResultsFaker;
import io.github.finoid.maven.plugins.codequality.fixtures.StepResultFaker;
import io.github.finoid.maven.plugins.codequality.fixtures.UnitTest;
import io.github.finoid.maven.plugins.codequality.fixtures.ViolationFaker;
import io.github.finoid.maven.plugins.codequality.report.Severity;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.List;

class StepResultsUnitTest extends UnitTest {
    @Test
    void givenNonPermissiveViolation_whenGetNonPermissiveViolations_thenReturnsExpectedViolation() {
        var nonPermissiveProjectStepResults = ProjectStepResultsFaker.projectStepResults()
            .withStepResult(StepResultFaker.stepResultFaker()
                .withIsPermissive(false)
                .withViolation(ViolationFaker.violation().create())
                .create())
            .create();

        var unit = StepResults.ofResults(List.of(nonPermissiveProjectStepResults));

        var result = unit.getNonPermissiveViolations(Severity.INFO);

        snapshot(result);
    }

    @Test
    void givenPermissiveViolation_whenGetNonPermissiveViolations_thenReturnsNoViolation() {
        var permissiveProjectStepResults = ProjectStepResultsFaker.projectStepResults()
            .withStepResult(StepResultFaker.stepResultFaker()
                .withIsPermissive(true)
                .withViolation(ViolationFaker.violation().create())
                .create())
            .create();

        var unit = StepResults.ofResults(List.of(permissiveProjectStepResults));

        var result = unit.getNonPermissiveViolations(Severity.INFO);

        Assertions.assertTrue(result.isEmpty());
    }

    @Test
    void givenNonPermissiveViolation_whenGetNonPermissiveViolationsAndHigherSeverity_thenReturnsNoViolation() {
        var nonPermissiveProjectStepResults = ProjectStepResultsFaker.projectStepResults()
            .withStepResult(StepResultFaker.stepResultFaker()
                .withIsPermissive(true)
                .withViolation(ViolationFaker.violation()
                    .withSeverity(Severity.INFO)
                    .create())
                .create())
            .create();

        var unit = StepResults.ofResults(List.of(nonPermissiveProjectStepResults));

        var result = unit.getNonPermissiveViolations(Severity.MINOR);

        Assertions.assertTrue(result.isEmpty());
    }
}