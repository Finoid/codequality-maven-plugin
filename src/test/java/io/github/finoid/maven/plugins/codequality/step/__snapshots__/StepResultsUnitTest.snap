io.github.finoid.maven.plugins.codequality.step.StepResultsUnitTest.givenNonPermissiveViolation_whenGetNonPermissiveViolations_thenReturnsExpectedViolation=[
  [
    {
      "columnNumber" : 17,
      "description" : "Description Zv2g7O0N4P",
      "fingerprint" : "LX8lvD1HROP1zV9O",
      "fullPath" : "/path/to/file/bNnRLyKE.java",
      "line" : 72,
      "relativePath" : "/file/KR1nHTx7.java",
      "rule" : "ZsDtD78bWT",
      "severity" : "MAJOR"
    }
  ]
]